### PR TITLE
feat: replace command with MicroBatchFramework

### PR DESCRIPTION
**Pull request by guitarrapc**
## TL;DR;

Try replace own command parser to [MicroBatchFramework](https://github.com/Cysharp/MicroBatchFramework).
Aim to establish simple command handling without thinking about args.

## Summary

Left : MicroBatchFramework / Right : Self command parser

image

## TODO

- [x] override default error
  - interceptor?
- [x] override default help message
  - seems like not the usecase, though? https://github.com/Cysharp/MicroBatchFramework/blob/master/src/MicroBatchFramework/CommandAttribute.cs#L15
- [x] Is any simple way to provide unix style help: -version --version -v / -help --help -h
  - should be require new Command Constructor.
- [x] Handle how to fallback to help when none of command match to args.

### CODE REVIEW

**Comment by guitarrapc** [comment]
`help` is caught before any interceptor.

> https://github.com/Cysharp/MicroBatchFramework/blob/f510c27a6a0394a8fe0a9f44d10384edb66479cb/src/MicroBatchFramework/BatchEngineHostBuilderExtensions.cs

I need filter args before passing to BatchEngine.

**Comment by guitarrapc** [comment]
interceptor could not override default error, just replace to `-help` argument.

**Comment by guitarrapc** [comment]
seems no way to provide unix argument style with simple way.

**Comment by guitarrapc** [comment]
Prepare `ArgsInterceptor(args)`  to override default MicroBatchFramework behavior. It's too sensitive as `help` and `list` are just framework's const.

**Comment by neuecc** [comment]
I've checked your codes and `0.4.5-beta9` now fit your usage, you can write more simplify.

```csharp
        static async Task Main(string[] args)
        {
             // not require ArgsInterceptor
            await new HostBuilder().RunBatchEngineAsync<Base64Batch>(args);
        }
```

```csharp
// multiple command and override default help command.
[Command(new[] { "help", "list", "-h", "-help", "--help" }, "show help")]
public void Help()
{    
    Console.WriteLine("Usage: base64urls [-version] [-help] [decode|encode|escape|unescape] [args]");
    Console.WriteLine("E.g., run this: base64urls decode QyMgaXMgYXdlc29tZQ==");
    Console.WriteLine("E.g., run this: base64urls encode \"C# is awesome.\"");
    Console.WriteLine("E.g., run this: base64urls escape \"This+is/goingto+escape==\"");
    Console.WriteLine("E.g., run this: base64urls unescape \"This-is_goingto-escape\"");
}
```

**Comment by guitarrapc** [comment]
great, let me check.

**Comment by guitarrapc** [comment]
@neuecc thanks, it works as expected.
BTW, is there any way to override default output when no arguments are passed?

image

image


Should I just handle before command evaluated?

```csharp
        private static readonly string[] validCommand = new[]
        {
            "help", "list", "-h", "-help", "--help",
            "version", "-v", "-version", "--version",
            "encode",
            "decode",
            "escape",
            "unescape",
        };
        static async Task Main(string[] args)
        {
            if (args == null || !args.Any() || !validCommand.Contains(args[0]))
            {
                args = new[] { "help" }.ToArray();
            }
            // TODO: How to fallback if none of arg is match to command. E.G., `base64urls version`
            await new HostBuilder().RunBatchEngineAsync<Base64Batch>(args);
        }
```

**Comment by neuecc** [comment]
currently, no argument was passed, shows help.
(maybe debug settings uses `"commandLineArgs": "invalid"` so is it no argument?)

But your latest commit(`feat: Fallback to help when none of command match to args.`) feature does not have.
Maybe I don't add it so your custom validCommand workaround is good.

**Comment by guitarrapc** [comment]
ah, yes you are right. ok, I understand your opinion.

* no arguments are passed will fall back to help by MicroBatchFramework.
* invalid command are passed will not fallback to help with MicroBatchFramework.
* I should handle this situation. ([latest commit](https://github.com/guitarrapc/Base64UrlCore/pull/2/commits/a3f60f3018b42a4637abbebb359d119ffb4dc7e8) is for it).

Thank you, I will release it as latest nuget global tools.

**Comment by guitarrapc** [comment]
invliad argument fall back to help is implemented with MicroBachFramework 0.4.5-beta10. It's perfect.

image
